dvc/output/__init__.py

```diff
@@ -1,136 +1,390 @@
-from dvc.output.local import OutputLOCAL
-from dvc.output.hdfs import OutputHDFS
-from dvc.output.ssh import OutputSSH
 
-from dvc.remote import Remote
-from dvc.remote.s3 import RemoteS3
-from dvc.remote.hdfs import RemoteHDFS
-from dvc.remote.local import RemoteLOCAL
-
-OUTS = [
-    OutputHDFS,
-    OutputS3,
-    OutputGS,
-    OutputSSH,
-    # NOTE: OutputLOCAL is the default choice
-]
-
-OUTS_MAP = {
-    Schemes.HDFS: OutputHDFS,
-    Schemes.S3: OutputS3,
-    Schemes.GS: OutputGS,
-    Schemes.SSH: OutputSSH,
-    Schemes.LOCAL: OutputLOCAL,
-}
-
-# NOTE: currently there are only 3 possible checksum names:
-#
-#    1) md5 (LOCAL, SSH, GS);
-#    2) etag (S3);
-#    3) checksum (HDFS);
-#
-# so when a few types of outputs share the same name, we only need
-# specify it once.
-CHECKSUM_SCHEMA = {
-    schema.Optional(RemoteLOCAL.PARAM_CHECKSUM): schema.Or(str, None),
-    schema.Optional(RemoteS3.PARAM_CHECKSUM): schema.Or(str, None),
-    schema.Optional(RemoteHDFS.PARAM_CHECKSUM): schema.Or(str, None),
-}
-
-TAGS_SCHEMA = {schema.Optional(str): CHECKSUM_SCHEMA}
-
-SCHEMA = CHECKSUM_SCHEMA.copy()
-SCHEMA[OutputBase.PARAM_PATH] = str
-SCHEMA[schema.Optional(OutputBase.PARAM_CACHE)] = bool
-SCHEMA[schema.Optional(OutputBase.PARAM_METRIC)] = OutputBase.METRIC_SCHEMA
-SCHEMA[schema.Optional(OutputBase.PARAM_TAGS)] = TAGS_SCHEMA
-SCHEMA[schema.Optional(OutputBase.PARAM_PERSIST)] = bool
-
-
-def _get(stage, p, info, cache, metric, persist=False, tags=None):
-    parsed = urlparse(p)
-
-    if parsed.scheme == "remote":
-        settings = stage.repo.config.get_remote_settings(parsed.netloc)
-        remote = Remote(stage.repo, settings)
-        return OUTS_MAP[remote.scheme](
-            stage,
-            p,
-            info,
-            cache=cache,
-            remote=remote,
-            metric=metric,
-            persist=persist,
-            tags=tags,
+logger = logging.getLogger(__name__)
+
+
+class DoesNotExistError(DvcException):
+    def __init__(self, obj):
+        name = "dependency" if obj.dependency else "output"
+        msg = "{} '{}' does not exist".format(name, obj)
+        super(DoesNotExistError, self).__init__(msg)
+
+
+class IsNotFileOrDirError(DvcException):
+    def __init__(self, obj):
+        name = "dependency" if obj.dependency else "output"
+        msg = "{} '{}' is not a file or directory".format(name, obj)
+        super(IsNotFileOrDirError, self).__init__(msg)
+
+
+class OutputAlreadyTrackedError(DvcException):
+    def __init__(self, path):
+        msg = "output '{}' is already tracked by scm (e.g. git)".format(path)
+        super(OutputAlreadyTrackedError, self).__init__(msg)
+
+
+class Output(object):
+    REMOTE = Remote
+
+    PARAM_PATH = "path"
+    PARAM_CACHE = "cache"
+    PARAM_METRIC = "metric"
+    PARAM_METRIC_TYPE = "type"
+    PARAM_METRIC_XPATH = "xpath"
+    PARAM_PERSIST = "persist"
+
+    PARAM_TAGS = "tags"
+
+    def __init__(self, stage, info, remote=None, dependency=False):
```

Comment:
there is regular mechanism to pass dicts - **kwargs.

Reply:
Fixed.